package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,690 @@
+
+        self.grid.fill_grid(self.coords_bbox)
+
+        self.prepared = True
+
+    def search(self, search_ids=None):
+        cdef ns_int i, j, size_search
+        cdef ns_int d, m
+        cdef NSResults results
+        cdef ns_int size = self.coords_bbox.shape[0]
+
+        cdef ns_int current_beadid, bid
+        cdef rvec current_coords
+
+        cdef ns_int cellindex, cellindex_adjacent, cellindex_probe
+        cdef ivec cellxyz, debug_cellxyz
+
+        cdef real[:, ::1] search_coords
+        cdef ns_int[:] search_ids_view
+
+        cdef ns_int xi, yi, zi
+        cdef real d2
+        cdef rvec shifted_coord, probe, dx
+
+        cdef ns_int nchecked = 0
+        cdef ns_int[:] checked = np.zeros(size, dtype=np.int)
+
+        cdef real cutoff2 = self.cutoff * self.cutoff
+        cdef ns_int npairs = 0
+
+        if not self.prepared:
+            self.prepare()
+
+        if search_ids is None:
+            search_ids=np.arange(size)
+        elif type(search_ids) == np.int:
+            search_ids = np.array([search_ids,], dtype=np.int)
+        elif type(search_ids) != np.ndarray:
+            search_ids = np.array(search_ids, dtype=np.int)
+
+        search_ids_view = search_ids
+        size_search = search_ids.shape[0]
+
+        results = NSResults(self.cutoff, self.coords, search_ids, self.debug)
+
+        cdef bint memory_error = False
+
+        with nogil:
+            for i in range(size_search):
+                if memory_error:
+                    break
+                current_beadid = search_ids_view[i]
+                cellindex = self.grid.cellids[current_beadid]
+                self.grid.cellid2cellxyz(cellindex, cellxyz)
+                for xi in range(DIM):
+                    if memory_error:
+                        break
+                    for yi in range(DIM):
+                        if memory_error:
+                            break
+                        for zi in range(DIM):
+                            if memory_error:
+                                break
+                            # Calculate and/or reinitialize shifted coordinates
+                            shifted_coord[XX] = self.coords[current_beadid, XX] + (xi - 1) * self.grid.cellsize[XX]
+                            shifted_coord[YY] = self.coords[current_beadid, YY] + (yi - 1) * self.grid.cellsize[YY]
+                            shifted_coord[ZZ] = self.coords[current_beadid, ZZ] + (zi - 1) * self.grid.cellsize[ZZ]
+                            probe[XX] = self.coords[current_beadid, XX] + (xi - 1) * self.cutoff
+                            probe[YY] = self.coords[current_beadid, YY] + (yi - 1) * self.cutoff
+                            probe[ZZ] = self.coords[current_beadid, ZZ] + (zi - 1) * self.cutoff
+                            # Make sure the shifted coordinates is inside the brick-shaped box
+                            for m in range(DIM - 1, -1, -1):
+                                while shifted_coord[m] < 0:

```

Comment:
I think it would be better if ``order`` can be used here in any way. It would be helpful for searching pairs i.e we just need to evaluate half the pairs i.e. 27/2 cells to get all the pairs (as every mirror pair will be included exactly once in those half-searches).